src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,152 @@
+        svclist = self.get_service(fs_name)
+        if svclist is None or len(svclist) == 0:
+            self.log.warn("Failed to fetch MDS service details for fs '{}'"
+                          .format(fs_name))
+            return None
+
+        svc = svclist[0]
+
+        # no need to update placement count if the current placement count is
+        # the required count
+        if svc.spec.placement.count == abscount:
+            return None
+
+        ps = copy.deepcopy(svc.spec.placement)
+        ps.count = abscount
+        newspec = ServiceSpec(service_type=svc.spec.service_type,
+                              service_id=svc.spec.service_id,
+                              placement=ps)
+        return newspec
+
+    def get_required_standby_count(self, fs_name: str) -> Optional[int]:
+        total = 0
+        assert self.fs_map is not None
+        for fs in self.fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs['mdsmap'].get('standby_count_wanted')
+        return total
+
+    def get_current_standby_count(self, fs_name: str, daemons: Optional[List[orchestrator.DaemonDescription]] = None) -> int:
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert self.fs_map is not None
+        total = 0
+        if daemons is None:
+            return total
+        daemon_names = {
+            d.name() for d in daemons
+        }
+        for sb in self.fs_map['standbys']:
+            full_name = "mds.{}".format(sb['name'])
+            if full_name in daemon_names:
+                total += 1
+        return total
+
+    def get_current_active_count(self, fs_name: str):
+        assert self.fs_map is not None
+        for fsys in self.fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return len(fsys.get('mdsmap').get('in'))
+        return 0
+
+    def verify_and_manage_mds_instance(self, fs_name: str):
+        assert self.fs_map is not None
+
+        daemons = self.get_daemons(fs_name)
+        standbys_required = self.get_required_standby_count(fs_name)
+        standbys_current = self.get_current_standby_count(fs_name, daemons)
+        active = self.get_current_active_count(fs_name)
+
+        self.log.info("standbys_required:{0}, standbys_current:{1}"
+                      .format(standbys_required, standbys_current))
+        if standbys_required is None:
+            return
+        total = standbys_current + active
+        if standbys_current > standbys_required:
+            # remove one mds at a time and wait for updated fs_map
+            self.log.info("killing {} standby mds ..."
+                          .format((standbys_current - standbys_required)))
+            total -= standbys_current - standbys_required
+        elif standbys_current < standbys_required:
```

Comment:
What is `max_mds` changes? I don't think this catches that case. If `standby_count_wanted` is 0 and `max_mds` is increased, then no new MDS is spawned.

Reply:
yeah, no new mds is spawned
I'll update the commit message and remove the reference to max_mds

Reply:
That's one of the most important functions of this module! Support for max_mds changes is not optional.

Reply:
What I meant was that if `max_mds` is bumped up, the standby count will drop as some standbys will be promoted to active mds and this condition will be noted by the plugin. The plugin will continue as usual to calculate the number of required standbys and spawn them accordingly.

But this needs some more work to incorporate max_mds changes. Thanks for poking.

Reply:
if `standby_count_wanted` is 0, then `standby_required` will be 0. But `max_mds` will never be satisfied.